[REEF-874] return application status after job is submitted

jwang98052 opened the pull request:
*This PR added  GetJobStatus() in IREEFClient and Yarn implementations
*Added named parameter YarnRmWebappHttpsAddressPropertyName for YarnConfigurationUrlProvider so tha tclient can configure it with the values set for the env

JIRA: [REEF-874](https://issues.apache.org/jira/browse/REEF-874)

This closes #

markusweimer commented on lang/cs/Org.Apache.REEF.Client/API/ApplicationState.cs:
```diff
@@ -0,0 +1,42 @@
+
+namespace Org.Apache.REEF.Client.API
+{
+    public enum ApplicationState
```
Please document.

jwang98052 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.Client/API/IREEFClient.cs:
```diff
@@ -43,5 +44,12 @@ public interface IREEFClient
+        /// </summary>
+        /// <returns></returns>
+        [Unstable("0.14", "Working in progress for rest API status returned")]
+        Task<ApplicationState> GetJobStatus();
```
For which job are we getting the status? Until now, `IREEFClient` supported multiple submissions. Wouldn't it be easier to return this as part of one of the `Submit...` methods?

jwang98052 replied:
Changed the API to pass application id

jwang98052 replied:
Job status needs to keep pulling after the Submit() is returned.. Therefore we need to separate API for it.

markusweimer commented on lang/cs/Org.Apache.REEF.Client/Common/HttpClientHelper.cs:
```diff
@@ -90,7 +91,12 @@ internal static List<string> GetRmUri(string filePath)
+        /// </summary>
+        /// <param name="filePath"></param>
+        /// <returns></returns>
+        public string GetAppId(string filePath)
```
Does this have to be `public`? It seems like a hack, ahm, implementation detail we won't want to support long term.

jwang98052 replied:
May be not right now. Long term, once the clients get app id, they can use it to call Rest API.  I changed it back to internal for now until we need I.

markusweimer commented on lang/cs/Org.Apache.REEF.Client/Common/HttpClientHelper.cs:
```diff
@@ -90,7 +91,12 @@ internal static List<string> GetRmUri(string filePath)
 
-        internal static string GetAppId(string filePath)
+        /// <summary>
+        /// Return teh applicaiton id of the Job
```
Typo.

jwang98052 replied:
Done

markusweimer commented on lang/cs/Org.Apache.REEF.Client/Common/HttpClientHelper.cs:
```diff
@@ -99,6 +105,11 @@ internal static string GetAppId(string filePath)
+        /// <summary>
+        /// Get and set application state
+        /// </summary>
+        public ApplicationState AppStatus { get; set; }
```
Why is the `public` and mutable? That would allow anyone with a handle on this object to make arbitrary changes.

jwang98052 replied:
Changed it to get only.

markusweimer commented on lang/cs/Org.Apache.REEF.Client/Common/IDriverHttpEndpoint.cs:
```diff
@@ -24,8 +25,28 @@ namespace Org.Apache.REEF.Client.Common
+        /// <summary>
+        /// Allow to set and get Application state
+        /// </summary>
+        ApplicationState AppStatus { get; set; }
```
Why is this `mutable`? Can the client change the state to `KILLED` and therefore kill the Driver?

jwang98052 replied:
Changed.

jwang98052 replied:
Long term, the status should be able to be set in driver as driver knows the progress. For client's killing request, we might need to do it through http server.

markusweimer commented on lang/cs/Org.Apache.REEF.Client/Local/LocalClient.cs:
```diff
@@ -135,10 +140,22 @@ public IDriverHttpEndpoint SubmitAndGetDriverUrl(IJobSubmission jobSubmission)
+        [Unstable("0.14", "Working in progress for rest API status returned")]
+        public async Task<ApplicationState> GetJobStatus()
+        {
+            await Task.Delay(1000); //1 seconds delay
```
This seems very dangerous. How did we choose 1s?

jwang98052 replied:
We don't really support status for local run time yet. This s a fake one. The API is defined in the interface, YarnREEFClient has a proper implementation because it can get status from Yarn Rest API. For local, we need to come out a proper implementation. I will create a Jira to follow it up.

markusweimer commented on lang/cs/Org.Apache.REEF.Client/YARN/Parameters/ConfigurationUriParameters.cs:
```diff
@@ -0,0 +1,31 @@
+
+namespace Org.Apache.REEF.Client.YARN.Parameters
+{
+    [NamedParameter("Yarn Rm Webapp Https Address PropertyName", defaultValue: "yarn.resourcemanager.webapp.https.address")]
```
Aren't these constants?

jwang98052 replied:
Those are different per environment.

markusweimer replied:
Shouldn't they then be fields of an interface that is bound per environment?

jwang98052 replied:
Currently it is defined as named parameter in YarnConfigurationUrlProvider so that client can configure it based on the environment. 
They are defined in a xml file of Yarn. They are keys, not values in the xml file. Otherwise, we would want to read it from the xml file on the fly. 
Anupam and I discussed about it and come out this solution for now. Rest API work is still in progress.
I have created REEF-868 to follow up the issue.

markusweimer commented on lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEF.cs:
```diff
@@ -63,7 +67,25 @@ private void Run()
+                result = status.Result;
+            }
+
+            Console.WriteLine("Application status : " + result.ToString());
```
Log this instead.

jwang98052 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEF.cs:
```diff
@@ -63,7 +67,25 @@ private void Run()
-            _reefClient.SubmitAndGetDriverUrl(helloJobSubmission);
+            var driverHttpEndpoint = _reefClient.SubmitAndGetDriverUrl(helloJobSubmission);
+
+            driverHttpEndpoint.AppStatus = GetJobStatus();
```
Why would one have to do this?

jwang98052 replied:
updated. 

Pulling status is just an example for the client.

markusweimer commented:
I've done a pass and it left me confused. The API suggests that the application state is somehow mutable from the Client side. Also, the example code now sets it? Wouldn't the status of an application be read-only from the client? And not set by application code, but instead provided by REEF?

jwang98052 commented:
Review comments are addressed and the change checked in.

markusweimer commented on lang/cs/Org.Apache.REEF.Client/Common/HttpClientHelper.cs:
```diff
@@ -64,6 +69,19 @@ public string GetUrlResult(string url)
+        ///// <summary>
+        ///// Get application Id
+        ///// </summary>
+        //public string AppId { get { return _appId; } }
```
Please remove commented out code.

jwang98052 replied:
Done

markusweimer commented on lang/cs/Org.Apache.REEF.Client/Common/HttpClientHelper.cs:
```diff
@@ -90,12 +108,17 @@ internal static List<string> GetRmUri(string filePath)
             {
-                var appId = sr.ReadLine();                
-                return appId;
+                _appId = sr.ReadLine();
```
Why do we need to overwrite this field over and over again?

jwang98052 replied:
Those were original code. We should clean up to read the data only once and leave them in variables. They are fixed per job submission. We can either have a separate Jira to do it, or as it is Unstable, we might change it soon with rest API submission.

markusweimer replied:
The change here makes the return value an instance variable. That is not the original behavior. Now. reads to `_appId` will return different things depending on when they are called.

jwang98052 replied:
appId was written to the file in REEF Java code after the Job is submitted to RM. It is unique per Job Submit. I can change the code to read it only once instead of reading every time it is called.

markusweimer commented on lang/cs/Org.Apache.REEF.Client/Common/HttpClientHelper.cs:
```diff
@@ -64,6 +69,19 @@ public string GetUrlResult(string url)
+        /// <summary>
+        /// Get application status
+        /// </summary>
+        public State AppFinalStatus
```
Does this need to be `public`?

jwang98052 replied:
Yes, that is the entire change for. We need to allow the client to get Job status after the Job is submitted.

markusweimer commented on lang/cs/Org.Apache.REEF.Client/Common/IDriverHttpEndpoint.cs:
```diff
@@ -24,8 +25,21 @@ namespace Org.Apache.REEF.Client.Common
     public interface IDriverHttpEndpoint
     {
+        /// <summary>
+        /// Get Uri result
```
I don't understand. What is the "Uri result"?

jwang98052 replied:
This was existing code. It actually returns http response result for a given uri as a web request. We can revisit it to either rename or as it is an Unstable, may drop later.

markusweimer replied:
The comment is new. We should improve upon it beyond adding spaces to the method name :)

jwang98052 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.Client/Common/IDriverHttpEndpoint.cs:
```diff
@@ -24,8 +25,21 @@ namespace Org.Apache.REEF.Client.Common
         string DriverUrl { get; }
+
+        /// <summary>
+        /// Allow to set and get Application state
```
It is get only.

jwang98052 replied:
Done

markusweimer commented on lang/cs/Org.Apache.REEF.Client/Local/LocalClient.cs:
```diff
@@ -51,8 +54,9 @@ public sealed class LocalClient : IREEFClient
         private readonly int _numberOfEvaluators;
         private readonly string _runtimeFolder;
-        private string _driverUrl;
+        //private string _driverUrl;
```
Please remove commented code.

jwang98052 replied:
Done

markusweimer commented on lang/cs/Org.Apache.REEF.Client/Local/LocalClient.cs:
```diff
@@ -106,6 +111,7 @@ public void Submit(IJobSubmission jobSubmission)
                 javaParams.TcpPortRangeCount.ToString(),
                 javaParams.TcpPortRangeTryCount.ToString()
                 );
+            _jobStatus = State.SUBMITTED;
```
This assumes that this instance of `IREEFClient` is used to submit only one job. This isn't how the interface was initially documented, and I don't think we need to make this assumption now. If this method is called more than once, `_subStatus` will be all messed up.

jwang98052 replied:
You are right. We don't have simulation for getting Job status in Local runtime as provided by Yarn Rest API. This is something we need to put in TODO list. I have create REEF-889 for it.

markusweimer commented on lang/cs/Org.Apache.REEF.Client/Local/LocalClient.cs:
```diff
@@ -131,19 +137,34 @@ public IDriverHttpEndpoint SubmitAndGetDriverUrl(IJobSubmission jobSubmission)
+            return _jobStatus;
         }
 
+        //public string DriverUrl
```
Remove commented out code.

jwang98052 replied:
Done

markusweimer commented:
I've done another pass.

jwang98052 commented:
addressed review comments and pushed in the change.

markusweimer commented on lang/cs/Org.Apache.REEF.Client/API/IREEFClient.cs:
```diff
@@ -43,5 +45,13 @@ public interface IREEFClient
+        /// <param name="appId"></param>
+        /// <returns></returns>
+        [Unstable("0.14", "Working in progress for rest API status returned")]
+        Task<State> GetJobFinalStatus(string appId);
```
How would one obtain the `appId`? 

I still wonder why we can't just return the app id as part of `SubmitAndGetDriverUrl`?

jwang98052 replied:
I exposed AppId as a property in IDriverHttpEndpoint which is the return of SubmitAndGetDriverUrl. But based on your last comments, I removed it. The _appId as a local variable is still available in the implementation of IDriverHttpEndpoint object. It is easy to be exposed as public API. 

Now IDriverHttpEndpoint  exposes AppFinalStatus(), which knows appId and then calls IREEFClient.GetJobFinalStatus(string appId). 

Another way is still expose appId in  IDriverHttpEndpoint, client then can use it to call GetJobFinalStatus(string appId); from IREEFClient directly.

markusweimer commented:
I'm still confused by this API change. As far as I understand, the intent is to expose the current status of a Driver to the client. This Driver may still be running or already shut down, which complicates matters. 

A straight forward way to do this to me is this:

  * Change `SubmitAndGetDriverUrl` to `SubmitAndGetJobStatus` which returns an `IDriverHandle`
  * Add `IDriverHandle GetDriverHandle(string appId)` to `IREEFClient`.

The `IDriverHandle` represents the running or completed Driver. It has the following structure:

```
public interface IDriverHandle{
  // Can return `null` when the Driver is no longer running.
  Uri DriverUri(get;);
  string AppId{get;}
  state State{get;}
}
```

That way, we can unify the APIs for when the job is running with the APIs for when it is done. In the latter case, a user can call `IREEFClient.GetDriverHandle` to get the status of the job.

Also, this might get us closer to the API @anupam128 is working on.

jwang98052 commented:
I agree the name change, make them readable: 
Change SubmitAndGetDriverUrl to SubmitAndGetJobStatus
change IDriverHttpEndpoint to IDriverHandle
SubmitAndGetJobStatus returns IDriverHandle just like SubmitAndGetDriverUrl  returns IDriverHttpEndpoint 

Inside IDriverHandle, DriverUri and AppId are unchanged once the Job is submitted. Client gets IDriverHandle  after calling SubmitAndGetDriverUrl. I am not sure if we want to expose IDriverHandle in IREEFClient for the caller of the REEFClient to get the handle again. Because what the client really needs to call again is GetStatus,as it may change while the Job is executing.

jwang98052 commented:
The code is modified to address a RM issue by adding MultipleRMUrlProvider. This also removed the dependency on the named parameter for the  yarn web address keys. 
I changed some of the names, e.g. SubmitAndGetDriverUrl to SubmitAndGetJobStatus. 
For IDriverHttpEndpoint, as it represents something returned from submit, so I named it as IJobSubmissionResult. 
For the code inside JobSubmissionResultImpl, originally named HttpEndPoitHelper, it needs some clean up. I am not sure if we should do it in a separate Jira or wait until rest API is done. 
The change so far has pushed in.

anupam128 commented on lang/cs/Org.Apache.REEF.Client/API/IREEFClient.cs:
```diff
@@ -37,11 +39,19 @@ public interface IREEFClient
+        /// <returns>IJobSubmissionResult</returns>
         [Unstable("0.13", "Working in progress for what to return after submit")]
-        IDriverHttpEndpoint SubmitAndGetDriverUrl(IJobSubmission jobSubmission);
+        IJobSubmissionResult SubmitAndGetJobStatus(IJobSubmission jobSubmission);
```
If this API is submit and GetJobStatus, do we need separate GetJobStatus?

jwang98052 replied:
This API is to return the result right after the submission. Status needs to keep pulling after it.

anupam128 replied:
Status right after submission is highly likely to be useless. What is the point of this API then?

jwang98052 replied:
We need to return an object that contains submission result. Some properties in it are available after submission and some like status need to keep pulling. Maybe the method name need to be changed. We mark is as unstable anyway.

anupam128 commented on lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/DataModel/ApplicationState.cs:
```diff
@@ -35,6 +35,10 @@ internal enum State
+
+        SUCCEEDED,
+
         FINISHED,
```
This enum does not have these values according to the documentation. Why are we adding these?
http://hadoop.apache.org/docs/r2.6.1/hadoop-yarn/hadoop-yarn-site/ResourceManagerRest.html

State 	The application state according to the ResourceManager - valid values are members of the YarnApplicationState enum: NEW, NEW_SAVING, SUBMITTED, ACCEPTED, RUNNING, FINISHED, FAILED, KILLED

Also see: https://github.com/apache/hadoop/blob/trunk/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-api/src/main/java/org/apache/hadoop/yarn/api/records/YarnApplicationState.java

anupam128 commented on lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/IRestClientFactory.cs:
```diff
@@ -42,7 +42,8 @@ public IRestClient CreateRestClient(Uri baseUri)
-                FollowRedirects = true
+                FollowRedirects = true,
+                MaxRedirects = 10
             };
```
Please revert this change. This I believe was only for debugging.

anupam128 commented on lang/cs/Org.Apache.REEF.Client/YARN/YARNREEFClient.cs:
```diff
@@ -117,10 +141,10 @@ private void Launch(IJobSubmission jobSubmission, string driverFolderPath)
+        //{
+        //    get { return _driverUrl; }
+        //}
 
```
Uncomment or remove?

jwang98052 replied:
Done

anupam128 commented on lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEF.cs:
```diff
@@ -63,7 +72,13 @@ private void Run()
+            var state = _reefClient.GetJobFinalStatus(appId);
+
+            Logger.Log(Level.Info, string.Format(CultureInfo.CurrentCulture,
+              "Job is submited with id: {0} and state: {1}", appId, state));
```
Do we actually intend to change HelloREEF? Seems unnecessary bloat for HelloWorld example.

jwang98052 replied:
I have removed most of my test code for running HelloREEF on yarn. This is just couple of lines to show user how to get status..

markusweimer replied:
Please remove this. The current APIs to get status aren't finalized and adding them to an example will just create confusion once we implement the real APIs.

anupam128 commented on lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestSimpleEventHandlers.cs:
```diff
@@ -48,7 +48,7 @@ public void TestSetup()
         public void RunSimpleEventHandlerOnLocalRuntime()
         {
-            string testFolder = DefaultRuntimeFolder + TestNumber++;
+            string testFolder = DefaultRuntimeFolder + Guid.NewGuid().ToString("N").Substring(4);
```
Why substring?

jwang98052 replied:
Just don't want too long folder name. We also clean up after the test. The opportunity of duplication is already very low.

markusweimer replied:
What's the reason for the desire for short folder names? Isn't it safer to use the long ones?

anupam128 commented:
I've done a pass.

jwang98052 commented:
Review comments have been addressed and Changes have been pushed in.

jwang98052 commented:
@markusweimer @anupam128 I have updated the PR to address all review comments and merged from latest master. Could you please review again?

anupam128 commented on lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/MultipleRMUrlProvider.cs:
```diff
@@ -0,0 +1,74 @@
+
+namespace Org.Apache.REEF.Client.YARN.RestClient
+{
+    public class MultipleRMUrlProvider : IUrlProvider
```
Please add unittests.

jwang98052 replied:
added

anupam128 commented on lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/YarnClient.cs:
```diff
@@ -130,19 +134,49 @@ private RestRequest CreateRestRequest(string resourcePath, Method method, string
+            }
+
+            throw new AggregateException("Failed Rest Request", exceptions);
         }
```
Add unittest for this new logic

jwang98052 replied:
added

anupam128 commented:
I have taken a look.

jwang98052 commented:
Change pushed in. 
@markusweimer can you review/merge it?

markusweimer commented on lang/cs/Org.Apache.REEF.Bridge/Org.Apache.REEF.Bridge.vcxproj.user:
```diff
@@ -0,0 +1,4 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Please remove this file, it is user-level VS configuration.

jwang98052 replied:
Good catch, accidently get it.

markusweimer commented on lang/cs/Org.Apache.REEF.Client/Common/JobSubmissionResultImpl.cs:
```diff
@@ -43,20 +45,57 @@ internal class HttpClientHelper : IDriverHttpEndpoint
             };
             _client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue(AppJson));
+
+            if (yarn)
```
This is a very dangerous design, as it enumerates all possible runtimes in a way that will be hard to find if we ever want to support e.g. Mesos in the C# code. A better way would be to add a new interface for this that the runtime in question needs to implement & bind. That way, new runtime authors have clear(er) guidance on what they need to do.

jwang98052 replied:
Totally agree. I was trying not to make too dramatic changes on the original code as it will change. For now, I introduced an abstract method GetDriverUri() at base level and allow each runtime to implement its own.

markusweimer commented on lang/cs/Org.Apache.REEF.Client/Common/JobSubmissionResultImpl.cs:
```diff
@@ -43,20 +45,57 @@ internal class HttpClientHelper : IDriverHttpEndpoint
+        }
 
+        /// <summary>
+        /// Get application final status from Yarn
```
So, this only works on YARN? How do we communicate this to the user of this API?

jwang98052 replied:
Updated unstable comments for it

markusweimer commented on lang/cs/Org.Apache.REEF.Client/Common/JobSubmissionResultImpl.cs:
```diff
@@ -120,29 +150,28 @@ internal static string GetTrackingUrl(string filePath)
         {
-            _driverUrl = GetTrackingUrl(filePath);
-            return _driverUrl;
+            return GetTrackingUrlAppId(filePath);
```
This opens a file written by the Java side of the client, right? How do we make sure that the file exists at this time? What if it wasn't written yet?

jwang98052 replied:
Yes, the file is written in Java code. I added checking anyway.

markusweimer commented on lang/cs/Org.Apache.REEF.Client/Common/JobSubmissionResultImpl.cs:
```diff
@@ -120,29 +150,28 @@ internal static string GetTrackingUrl(string filePath)
+                    fullDriverUrl = "http://" + driverUrl + "/";
                     break;
                 }
                 Thread.Sleep(1000);
```
Why the sleep? I'm very hesitant to add real time units anywhere to the code, as they will create bugs in the future.

jwang98052 replied:
In this case, we are waiting for the data to be written to the file. If it is not ready, we need to wait for sometime and then try again. We don't want to keep trying and using CPU time.

markusweimer commented on lang/cs/Org.Apache.REEF.Client/Local/LocalClient.cs:
```diff
@@ -132,16 +134,23 @@ public IDriverHttpEndpoint SubmitAndGetDriverUrl(IJobSubmission jobSubmission)
 
-            Logger.Log(Level.Info, "Submitted the Driver for execution. Returned driverUrl is: " + _driverUrl);
-            return helper;
+            Logger.Log(Level.Info, string.Format(CultureInfo.CurrentCulture,
```
Please pass the format string straight to the logger. That way, it won't be created unless needed.

jwang98052 replied:
It is already using string.Format(). Can you elaborate?

markusweimer commented on lang/cs/Org.Apache.REEF.Client/Local/LocalClient.cs:
```diff
@@ -132,16 +134,23 @@ public IDriverHttpEndpoint SubmitAndGetDriverUrl(IJobSubmission jobSubmission)
+        public async Task<FinalState> GetJobFinalStatus(string appId)
         {
-            get { return _driverUrl; }
+            await Task.Delay(1000); //1 seconds delay
```
Why the real time units here?

jwang98052 replied:
This is just a place holder as we have to implement the interface which match the async syntax. It is TODO. I changed the delay to 0.

markusweimer commented on lang/cs/Org.Apache.REEF.Client/YARN/YARNREEFClient.cs:
```diff
@@ -83,11 +90,26 @@ public IDriverHttpEndpoint SubmitAndGetDriverUrl(IJobSubmission jobSubmission)
 
-            var httpClient = new HttpClientHelper();
-            _driverUrl = httpClient.GetDriverUrlForYarn(pointerFileName);
+            Logger.Log(Level.Info, string.Format(CultureInfo.CurrentCulture,
```
Same as above: Please pass the format string to the logger instead of forcing its expansion here.

jwang98052 replied:
Done.

markusweimer commented:
@jwang98052 I've done a pass.

@anupam128 How close are we with the real APIs? This touches on a lot of code in many ways we will likely have to undo once that lands. If that is just around the corner, it might make sense not to merge this at all.

jwang98052 commented:
I have addressed review comments and pushed in the change. 

@markusweimer @anupam128 
This branch now is used in many places in our other effort cross different source depots. It is very easy to be out of sync and miss something if it is kept in a branch. I would strongly think we should merge it to master to bring the effort to current. I have labeled all the necessary places as [unstable] and put TODO with Jira numbers. Checking it in would force any future changes to ensure regression testing pass and the scenarios we support today should be still supported.

Even after Rest API is ready, some of the code may still be needed, such ad tracking URL for local run time, pull url result, etc. Some of the classes we added in this PR will remain. For example, MultipleRMUrlProvider is just another provider and it is useful and generic. FinalState would remain as it is needed for Rest API as well.

markusweimer commented:
OK, let's move forward with this. Before we can merge this, we need to resolve two issues:

  1. The code doesn't rebase cleanly onto current `master`. Julia, can you clean that up?
  2. We need to test .NET code on both local and YARN

jwang98052 commented:
I will follow up on those two. In fact, I have been keep testing it on local and yarn before every push. I will do it again after rebase.

jwang98052 commented:
Cheery picked up on top of the master and force pushed to the branch. 
Tested on both local and Yarn. 
@markusweimer could you please merge for me before more conflicts come?

markusweimer commented:
I will test and merge.
